content/docs/user-guide/gems/reference/kythera-ai/spatial-query-system.md

```diff
@@ -0,0 +1,723 @@
+---
+linkTitle: Spatial Query System
+title: Spatial Query System
+description: Overview of the Kythera AI Spatial Query System
+weight: 1000
+toc: true
+---
+
+![Spatial queries as shown by debug draw](/images/user-guide/gems/kythera-ai/spatial-query-system.png)
+
+Finding positions in a game world is a common requirement. Usually these are positions for agents to move to; there are many other purposes, including spawning and choosing points to attack. During development, the requirements for choosing these positions can become complex, making performance crucial. Kythera's Spatial Query System (SQS) is a powerful, general-purpose API for describing and efficiently processing these location-based queries.
+
+            
+
+When an agent needs to choose a position to move to, it can do so by executing a Spatial Query which will return the best position for a given set of criteria. The query itself is described in a specialized query language, comprising:
+
+*   **Generators**, which produce a set of positions to consider
+    
+*   **Conditions**, which test each position and exclude inappropriate ones
+    
+*   **Scores**, which weight and rank the appropriate positions in order to find the best candidate
+    
+
+A query is constructed once, then executed any number of times by the Spatial Query System. The system handles queuing, time-slicing and asynchronous processing of multiple queries over a number of frames. When a query is completed, it uses the Kythera signal mechanism to communicate the result back to the original caller.
+
+Spatial queries are specified in text files, which are parsed and interpreted at run-time. This allows rapid iteration, in particular for high-level agent-based queries, as the files can be reloaded while the game is running.
+        
+
+## Writing spatial queries
+
+Each level in a game can include a file called `SpatialQueries.sqs` which defines a collection of Spatial Queries. Queries common to multiple levels can be placed in the global Scripts folder. Whenever a level is loaded, all queries are first unloaded and then the global queries are loaded in. The level-specific queries come next and will overwrite any global queries with the same name.
+
+There is a console command, `kyt_ReloadSpatialQueries`, which will reload the queries in the file on-demand, allowing rapid iteration during development.
+In addition to the queries written in C++ as shown above, queries can also be provided in a file. These files will be parsed by the Spatial Query System, and used to construct spatial queries. The syntax for these queries is the same as for C++ queries.
+
+
+### Query structure
+
+A spatial query consists of four blocks - `Generators`, `Conditions`, `Scores` and `Fallbacks`. When the query is executed:
+
+*   The Generators are run in parallel (via time-slicing) and each returns one or more candidate positions or entities (such as cover points) as output
```

Comment:
If this is a guaranteed ordering of steps, would recommend using a numbered list.